File: zemiao/25.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        def length(head):
          i = 0
          while head:
            i += 1
            head = head.next
          return i
        
        '''Reverses [head:head+len],
        returns (head,tail)'''
        def reverse(head, len):
          if not head:
            return None, None
          if length(head) < len:
            return head, None
          tail, newHead, next = head, head, head.next
          i = 1
          while next and i<len:
            i += 1
            newNext = next.next
            next.next = newHead
            newHead = next
            next = newNext
          tail.next = next
          return newHead, tail
        
        dummy = ListNode(0)
        dummy.next = head
        tail = dummy
        while tail:
          tail.next, tail = reverse(tail.next, k)
        return dummy.next
        

File: tnzhou/25.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    int getLength(ListNode* head) {
        int n = 0;
        ListNode* cur = head;
        while (cur != NULL) {
            n++;
            cur = cur->next;
        }
        return n;
    }
    
    void print(ListNode* head) {
        ListNode* cur = head;
        while (cur != NULL) {
            cout << cur->val;
            cur = cur->next;
        }
        cout << endl;
    }
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        len = getLength(head);
        ListNode* cur_fa = NULL;
        ListNode* cur = head;
        for (int i = 0; i < len / k; i++) {
            for (int movetimes = k - 1; movetimes > 0; movetimes--) {
                ListNode* itr = cur;
                ListNode* itr_fa = cur_fa;
                cout << cur->val << endl;
                for (int j = 0; j < movetimes; j++) {
                    if (itr_fa != NULL)
                        itr_fa->next = itr->next;
                    else {
                        head = itr->next;
                    }
                    if (itr == cur) {
                        cur = itr->next;
                    }
                    itr->next = itr->next->next;
                    if (itr_fa != NULL) {
                        itr_fa->next->next = itr;
                        itr_fa = itr_fa->next;
                    } else {
                        head->next = itr;
                        itr_fa = head;
                    }
                    // print(head);
                }
            }
            for (int j = 0; j < k; j++) {
                cur_fa = cur;
                cur = cur->next;
            }
        }
        return head;
    }
private:
    int len;
};
